File: CSharp-Part-2/1. Arrays/Arrays/BinarySearch/BinarySearch.cs
// Write a program that finds the index of given element in a sorted array of integers by using the Binary search algorithm.

using System;

class BinarySearch
{
    static void Main()
    {
        Console.WriteLine("Enter array length: ");
        int n = int.Parse(Console.ReadLine());
        Console.WriteLine("Enter a number to search for: ");
        int inputNumber = int.Parse(Console.ReadLine());
        Console.WriteLine("Enter array contents: ");
        int[] inputArray = new int[n];
        for (int i = 0; i < n; i++)
        {
            inputArray[i] = int.Parse(Console.ReadLine());
        }

        Array.Sort(inputArray);
        int index = binarySearch(inputArray, inputNumber, 0, n - 1);
        if (index != -1)
        {
            Console.WriteLine("The number {0} was found at index {1}.", inputNumber, index);
        }
        else
        {
            Console.WriteLine("The number {0} is not present in the array.", inputNumber);
        }
    }
    static int binarySearch(int[] inputArray, int inputNumber, int iMin, int iMax)
    {
        while (iMax >= iMin)
        {
            int midPoint = (iMin + iMax) / 2;
            if (inputArray[midPoint] == inputNumber)
            {
                return midPoint;
            }
            else if (inputArray[midPoint] < inputNumber)
            {
                iMin = midPoint + 1;
            }
            else
            {
                iMax = midPoint - 1;
            }
        }
        return -1;
    }
}

File: JavaScript-Fundamentals/07. Arrays/binary-search.js
// Write a script that finds the index of given element in a sorted array of integers by using the binary search algorithm.

var sampleArray = [13, 7, 2, 6, 3, 15, 1, 1, 8, 33, 9, 39, 81],
    numberToFind = 81;
    
function sortNumber(a, b) {
    return a - b;
}

function binarySearch(array, num, min, max) {
    var mid = min + Math.floor((max - min) / 2);
    
    if (max < min || num > array[max]) {
        return -1;
    }

    if (array[mid] > num) {
        return binarySearch(array, num, min, mid - 1);
    } else if (array[mid] < num) {
        return binarySearch(array, num, mid + 1, max);
    } else {
        return mid;
    }
}

sampleArray.sort(sortNumber);

console.log(binarySearch(sampleArray, numberToFind, 0, sampleArray.length - 1));